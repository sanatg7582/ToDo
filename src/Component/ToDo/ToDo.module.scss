.todoContainer {
  width: 100%;
  float: left;
  padding: 30px;
  .todoForm {
    width: 45%;
    float: left;
    .todoFormGroup{
      width: 50%;
    }
  }
  .todoList {
    width: 45%;
    float: left;
    table {
      width: 100%;
    }
    table,
    th,
    td {
      text-align: center;
    }
  }
}
