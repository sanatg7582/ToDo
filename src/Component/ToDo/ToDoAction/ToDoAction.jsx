import React from "react";
import PropTypes from "prop-types";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTrash, faPen } from "@fortawesome/free-solid-svg-icons";
import Button from "../../Shared/Button";
/**
 * This function return Edit and Delete button
 * @param {number} index
 * @param {func} handleOnClickRemove
 * @param {func} handleOnClickUpdate
 * @returns Button Element
 */
const ToDoAction = props => {
  const { index, handleOnClickRemove, handleOnClickUpdate } = props;
  return (
    <>
      <Button
        buttonType="button"
        btnClass="warning"
        handleOnClick={handleOnClickUpdate}
        index={index}
      >
        <FontAwesomeIcon icon={faPen} />
      </Button>
      &nbsp;
      <Button
        buttonType="button"
        btnClass="danger"
        handleOnClick={handleOnClickRemove}
        index={index}
      >
        <FontAwesomeIcon icon={faTrash} />
      </Button>
    </>
  );
};
// default props value
ToDoAction.defaultProps = {
  index: 0,
  handleOnClickRemove: () => {},
  handleOnClickUpdate: () => {}
};
//prop type validation
ToDoAction.propTypes = {
  index: PropTypes.number,
  handleOnClickRemove: PropTypes.func.isRequired,
  handleOnClickUpdate: PropTypes.func.isRequired
};
export default ToDoAction;
