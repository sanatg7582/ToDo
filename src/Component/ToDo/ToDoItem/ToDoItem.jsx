import React from "react";
import PropTypes from "prop-types";
import Text from "../../Shared/Text";
import ToDOStatus from "../ToDoStatus";
import ToDoCompleteStatus from "../ToDoCompleteStatus";
import ToDoAction from "../ToDoAction";
/**
 * This function for todo Item row in table
 * @param {number} index
 * @param {string} todoHeading
 * @param {bool} todoStatus
 * @param {bool} todoIsCompleteStatus
 * @param {func} updateToDoStatus
 * @param {func} completeToDoStatus
 * @param {func} removeToDo
 * @param {func} updateToDo
 * @returns tr element
 */
const ToDoItem = props => {
  const {
    index,
    todoHeading,
    todoStatus,
    todoIsCompleteStatus,
    updateToDoStatus,
    completeToDoStatus,
    removeToDo,
    updateToDo
  } = props;
  return (
    <tr>
      <td>
        <Text strong={true} heading={todoHeading} />
      </td>
      <td>
        <ToDOStatus
          todoStatus={todoStatus}
          handleOnClick={updateToDoStatus}
          index={index}
        />
      </td>
      <td>
        <ToDoCompleteStatus
          isCompleteStatus={todoIsCompleteStatus}
          handleOnClick={completeToDoStatus}
          index={index}
        />
      </td>
      <td>
        <ToDoAction
          index={index}
          handleOnClickRemove={removeToDo}
          handleOnClickUpdate={updateToDo}
        />
      </td>
    </tr>
  );
};
// default props value
ToDoItem.defaultProps = {
  index: 0,
  todoHeading: "",
  todoStatus: true,
  todoIsCompleteStatus: false,
  updateToDoStatus: () => {},
  completeToDoStatus: () => {},
  removeToDo: () => {},
  updateToDo: () => {}
};
//props type validation
ToDoItem.propTypes = {
  index: PropTypes.number,
  todoHeading: PropTypes.string.isRequired,
  todoStatus: PropTypes.bool.isRequired,
  todoIsCompleteStatus: PropTypes.bool.isRequired,
  updateToDoStatus: PropTypes.func.isRequired,
  completeToDoStatus: PropTypes.func.isRequired,
  removeToDo: PropTypes.func.isRequired,
  updateToDo: PropTypes.func.isRequired
};
export default ToDoItem;
