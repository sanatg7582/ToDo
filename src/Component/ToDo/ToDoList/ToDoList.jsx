import React, { useState } from "react";
import PropTypes from "prop-types";
import ToDoItem from "../ToDoItem";
import ToDoFilters from "../ToDoFilters";

/**
 * This component for render ToDoList
 * @param {array} todos // array of object
 * @param {func} updateToDos // func
 * @param {func} handleOnClick // func
 */

const ToDoList = (props) => {
  const [filterActive, setFilterActive] = useState(0);
  const [searchterm, setSearchterm] = useState();
  const { todos, updateToDos, handleOnClick } = props;

  // remove toDo
  const removeToDo = title => {
    updateToDos(
      todos.filter(itm => {
        return itm.todo !== title;
      })
    );
  };

  // Update toDo Status
  const updateToDoStatus = index => {
    updateToDos(
      todos.map((item, key) =>
        key === index ? { ...item, status: !item.status } : item
      )
    );
  };

  // Update complete ToDo status
  const completeToDoStatus = index => {
    updateToDos(
      todos.map((itm, key) =>
        key === index ? { ...itm, isCompleted: !itm.isCompleted } : itm
      )
    );
  };

  // update ToDo
  const updateToDo = (e, index) => {
    handleOnClick(e, index);
  };


  let newtoDos = [...todos];
  if (filterActive === 1) {
    newtoDos = todos.filter(itm => {
      return itm.status === true
    });
  }
  if (filterActive === 2) {
    newtoDos = todos.filter(itm => {
      return itm.isCompleted === true
    });
  }
  if (searchterm && searchterm.length > 0) {
    newtoDos = newtoDos.filter(itm => itm.todo.indexOf(searchterm) >= 0);
  }

  return (
    <>
      <ToDoFilters
        handleOnClick={(status) => { setFilterActive(status) }}
        filterActive={filterActive}
        searchterm={searchterm}
        setSearchtermfun={(key) => { setSearchterm(key) }}
      />

      <table>
        <thead>
          <tr>
            <th>ToDo</th>
            <th>Status</th>
            <th>IsCompleted</th>
            <th>Action</th>
          </tr>
        </thead>
        <tbody>
          {newtoDos.map((todo, index) => {
            return (
              <ToDoItem
                index={index}
                todoHeading={todo.todo}
                todoStatus={todo.status}
                todoIsCompleteStatus={todo.isCompleted}
                updateToDoStatus={() => updateToDoStatus(index)}
                completeToDoStatus={() => completeToDoStatus(index)}
                removeToDo={() => removeToDo(todo.todo)}
                updateToDo={updateToDo}
                key={index}
              />
            );

          })}
        </tbody>
      </table>
    </>
  );

}
// default props 
ToDoList.defaultProps = {
  todos: [],
  updateToDos: () => { },
  handleOnClick: () => { }
}
// PropTypes validation
ToDoList.propTypes = {
  todos: PropTypes.array,
  updateToDos: PropTypes.func,
  handleOnClick: PropTypes.func
};
export default ToDoList;
