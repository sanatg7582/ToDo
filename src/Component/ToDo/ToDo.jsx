import React, { useState } from "react";
import styles from "./ToDo.module.scss";
import Input from "../Shared/Input";
import Button from "../Shared/Button";
import ToDoList from "./ToDoList";

/**
 * This component for render ToDoList and ToDo form
 */
const ToDo = () => {
  const [index, setIndex] = useState(-1);
  const [todo, setTodo] = useState();
  const [todoData, setTodoData] = useState([
    { todo: "test", status: true, isCompleted: false }
  ]);

  const handleChange = e => {
    const { value } = e.target;
    setTodo(value);
  };

  const addToDo = e => {
    e.preventDefault();
    if (index !== -1) {
      // update To DO
      for (let i = 0; i < todoData.length; i++) {
        if (i === index) {
          todoData[i].todo = todo;
        }
      }
      setTodoData(todoData);
      setIndex(-1);
      setTodo();
    } else {
      // add To Do
      setTodoData([
        ...todoData,
        {
          todo: todo,
          status: true,
          isCompleted: false
        }
      ]);
      setIndex(-1);
      setTodo();
    }
  };

  // get toDo detail for edit
  const getToDoDetail = (e, index) => {
    setIndex(index);
    setTodo(todoData[index].todo);
  };

  return (
    <div className={styles.todoContainer}>
      <div className={styles.todoForm}>
        <form onSubmit={e => addToDo(e)}>
          <div className={styles.todoFormGroup}>
            <Input
              InputLabel="Todo"
              InputType="text"
              InputName="todo"
              IsRequired={true}
              handleChange={handleChange}
              InputValue={todo}
              InputMaxlength="30"
            />
          </div>
          <Button
            buttonType="submit"
            buttonName="Save ToDo"
            btnClass="success"
          />
        </form>

      </div>
      <div className={styles.todoList}>
        <ToDoList
          todos={todoData}
          handleOnClick={getToDoDetail}
          updateToDos={arr => {
            setTodoData(arr);
          }}
        />
      </div>
    </div>
  );
};
export default ToDo;
