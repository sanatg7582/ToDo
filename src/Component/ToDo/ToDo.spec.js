import React from "react";
import { mount } from "enzyme";
import ToDo from "./ToDo";
import ToDoItem from "./ToDoItem";
import Text from "../Shared/Text";
import ToDoList from "./ToDoList";
import ToDoAction from "./ToDoAction";
describe("Load <ToDo />", () => {
  let wrapper = "";
  beforeAll(() => {
    wrapper = mount(<ToDo />);
  });
  it("renders without crashing", () => {
    expect(wrapper).toBeTruthy();
  });

  it("Checking form length", () => {
    expect(wrapper.find("form")).toHaveLength(1);
  });
  it("todo form submit", () => {
    wrapper
      .find("form")
      .find("button")
      .simulate("submit");
    expect(wrapper.find(ToDoItem)).toHaveLength(2);


  });
  it("checking input props", () => {
    expect(
      wrapper
        .find("form")
        .find("input")
        .props().type
    ).toEqual("text");
    wrapper
      .find("form")
      .find("input")
      .simulate("change");
    expect(
      wrapper
        .find("form")
        .find("input")
        .props().value
    ).toBe("");
  });
  it("Checking toDo loop value ", () => {
    expect(
      wrapper
        .find(ToDoItem)
        .find(Text)
        .first()
        .find("span")
        .text()
    ).toEqual("test");
  });
  it("checking ToDo detail", () => {
    wrapper
      .find(ToDoList)
      .find(ToDoAction)
      .find("button")
      .first()
      .simulate("click");
    expect(
      wrapper
        .find("form")
        .find("input")
        .props().value
    ).toBe("test");
  });
  it("ds", () => {
    const mock = jest.fn().mockImplementation(() => "bar");
    expect(mock("foo")).toBe("bar");
    expect(mock).toHaveBeenCalledWith("foo");
  })
  it("SnapShot test", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
