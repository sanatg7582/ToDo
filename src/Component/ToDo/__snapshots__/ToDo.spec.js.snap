// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Load <ToDo /> SnapShot test 1`] = `ReactWrapper {}`;
